File: demos/02-havoc-coverage/input_param_modified/add_one_verified.cpp
#include "validator_const.h"

/*
DEMO: Verified (verifies) because the validator's pointer parameter is const.

The interface is declared in validator_const.h:
    virtual void validate(const uint32_t* val) = 0;

The validator is passed in as an IValidator* parameter, so its concrete
implementation is opaque to this translation unit (and to SAW). SAW
applies its vtable havoc model — but the const pointee tells the model
that the call cannot write through `val`. So `input` after the call
still equals `x`, and `input + 1 == x + 1` holds for all x.

In production code, the SAL annotation _In_ has the same effect:
    virtual void validate(_In_ uint32_t* val) = 0;
*/

// Spec: add_one(x) = x + 1
uint32_t add_one(IValidator* v, uint32_t x) {
    uint32_t input = x;

    // const pointer — SAW knows *input is unchanged
    v->validate(&input);

    // input is guaranteed to still equal x
    return input + 1;
}

File: tests/e2e/cases/02-havoc-coverage/input_param_modified/add_one_verified.rs
/*
DEMO: Verified — validator takes `&u32`, so it cannot write through the
pointer.

Mirrors add_one_verified.cpp,
which marks the validator parameter as `const uint32_t*`. In Rust the
equivalent is `&u32` — the borrow checker statically forbids the
callee from mutating the value. SAW sees that `input` is unchanged
across the call and concludes `input + 1 == x + 1`.
*/

fn validate(_val: &u32) {
    // Read-only borrow — cannot assign through `_val`.
}

fn add_one(x: u32) -> u32 {
    let input = x;
    validate(&input);
    input + 1
}
